HealthVault/Classes/Caching/Private/MHVThingCacheDatabase.m

```diff
@@ -63,7 +64,7 @@ - (instancetype)initWithKeychainService:(id<MHVKeychainServiceProtocol>)keychain
     return self;
 }
 
-- (NSError *_Nullable)deleteDatabase
+- (void)deleteDatabaseWithCompletion:(void (^)(NSError *_Nullable error))completion;
```

Comment:
I think that this method should be 'resetDatabaseWithCompletion:' and it should delete the current database and the code in 'setupDatabaseWithCompletion:' should be merged into this method. I'm still seeing potential threading issues with the code in it's current state, that we can resolve by simply making sure there is always a database. You can still use self.lockObject to synchronize, but instead of a self.deleted BOOL checking for the existence of a database, you have a synchronized method to check the status of the database. All code in the 'resetDataBaseWithCompletion:' is wrapped in @synchronized(self.lockObject). Now you can create a single method to check the database status:

 - (BOOL)isDatabaseReady
{
    @synchronized (self.lockObject)
    { 
          return self.persistentStoreCoordinator != nil && self.managedObjectContext != nil; 
    }
}